KEP-753: add termination alternatives for beta (#4183)
kubernetes/enhancements: keps/sig-node/753-sidecar-containers/README.md

## Summary

> <!--

Reviewer: Why are we bringing back comments?

   sidecar container terminates.
4. This continues until all sidecar containers have terminated, or the Pod's termination grace period
   expires.
5. In the latter case, all remaining containers are notified by a `SIGTERM`, followed by a fixed
   grace period of 2s and finally terminated.
6. The Pod will be terminated after that.

Pseudocode for the above:

```
func terminatePod() {
  // notify all sidecar containers with preStop hook, asynchronously
  for sidecar in sidecarContainers {
    if sidecar has preStop hook {
      go execute preStop hook // async
    }
  }
  // notify all containers with preStop hook and then SIGTERM, asynchronously
  for container in containers {
    if container has preStop hook {
      go func(container) { // async
        execute preStop hook
        send SIGTERM
      }
    }
  }
  for {
    switch {
      case grace period expired:
        for anyContainer in sidecarContainers + containers {
          if anyContainer is running {
            send SIGTERM
          }
        }
        sleep 2s
        for anyContainer in sidecarContainer + containers {
          if anyContainer is running {
            send SIGKILL
          }

>         }

Reviewer 2: I think we can simplify the code by introducing the `stopContainer`.

```go
// stopContainer is a remote call to the container runtime
// to stop the container.
function stopContainer(c container, tgps int) {
  // container runtime sends SIGTERM to the container
  // and waits for it to stop within the tgps.
  // If not killed, the container runtime will send SIGKILL to kill the container.
}
```

Author: maybe... does it still make it clear what's happening to all containers in case of TGPS expiration? because this was one of the main goals - I have the feeling it hides it by adding the call to function

Reviewer 2: Separating the one remote call (it is actually one remote call and kubernetes cannot control its behavior) into multiple commands makes it a bit complicated and inaccurate.
- regular containers should not get SIGTERM twice at L900.
- regular containers should not wait for extra 2s if preStop does not consume all TGPS at L903.

nit: If you think the detail is not that important, I am okay with that.

Author: fair enough for the 2s delay before SIGKILL

Reviewer 2: > please realize that there is a `if anyContainer is running {` before sending SIGTERM

Got it, you mean running = not terminating nor terminated.

Reviewer: IMO this psuedocode added SO MUCH clarity.  Thank you

        return
      case all containers are terminated:
        // sidecars are terminated in reverse order
        for sidecar in reverse(sidecarContainers) {
          // sidecar is already terminating, let it finish
          if sidecar is terminating {
            break
          }
          // next sidecar to terminate
          else if sidecar is running {
            send SIGTERM
            break
          }
        }

>         sleep 1s

Reviewer 2: Maybe we can just remove this sleep as this is just a pseudo code.

      case all sidecarContainers are terminated:
        return
      default:
        sleep 1s
    }
  }
}
```

It is worth noting that, like with regular containers, `PreStop` hook must complete before the `SIGTERM`
signal to stop the sidecar container can be sent. Therefore, ordering and graceful termination of sidecars
can only be guaranteed if the `PreStop` hook completes within the Pod's termination grace period.

Sidecars continue to be restarted until they enter the `Terminated` state which they are notified

> by a `SIGTERM` signal. This is to ensure that sidecars that fail are restarted until the TGPS expires.

Reviewer 3: Also say that the `preStop` will be called on restarted container

Author: yes it was implied

### Termination of containers

In Alpha sidecar containers will be terminated as regular containers. No special
or additional signals will be supported.

In Beta we have thought about introducing additional termination grace period fields
to manage termination duration
([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
and leverage these fields to add reverse order termination of sidecar containers
after the primary containers terminate.

> However, we decided on an alternative that don't require additional fields or hooks while keeping

Reviewer 4: ```suggestion
However, we decided on an alternative that doesn't require additional fields or hooks while keeping
```

> the desired behaviors when Pods with sidecars are terminated:

Reviewer 3: ```suggestion
the desired behaviors when Pods with sidecars are terminated. While original approach works better with truly graceful termination where consistency is more important than time taken, proposed approach works for that scenario as well as a more and more popular scenario of limited time to terminate when graceful termination is set by external requirement and Pods needs to do best to gracefully terminate as much as possible (think of a Spot Instances with 30 seconds notification).

Here is the proposed approach:
```

> 1. Sidecar containers that have a `PreStop` hook will be notified when the Pod is terminating

Reviewer 4: ```suggestion
1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
```

   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
   the Pod's termination grace period countdown.
2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
3. Once the last primary container terminates, the last started sidecar container is notified by

>    sending a `SIGTERM` signal.

Reviewer 3: ```suggestion
   sending a `SIGTERM` signal, even if `PreStop` hook hasn't completed yet.
```

4. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
   sidecar container terminates.
5. This continues until all sidecar containers have terminated, or the Pod's termination grace period
   expires. In the latter case, the Pod is force terminated and all sidecar containers are killed by

>    sending a `SIGKILL` signal.

Reviewer 2: This looks little bit confusing. A container will always get minimum grace period (2s) even if the termination grace period expires.

https://github.com/kubernetes/kubernetes/pull/120620

Reviewer 2: nit:

'In the latter case, all containers are terminated with minimum grace period and the Pod will be terminated after that.'

If we compare this to the initial proposal, the following behaviors are preserved:
- Sidecars should not begin termination until all primary containers have
  terminated.
- Sidecars should terminate serially and in reverse order. I.e. the first
  sidecar to initialize should be the last sidecar to terminate.

> 

Reviewer 3: ```suggestion

The additional benefits of this approach comparing to initial proposal:
- If graceful termination period is short, and mostly taken by the main container, the sidecar containers has more time to gracefully terminate, for example, clear up buffers of logging container.
- There is absolutely no change in behavior of main containers - they start graceful termination at exact same time as before and can utilize as much of the graceful termination period as they need. The Pod graceful termination period semantic also stay unchanged
```

### Termination of containers

In Alpha sidecar containers will be terminated as regular containers. No special
or additional signals will be supported.

In Beta we have thought about introducing additional termination grace period fields
to manage termination duration
([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
and leverage these fields to add reverse order termination of sidecar containers
after the primary containers terminate.

However, we decided on an alternative that doesn't require additional fields or hooks while keeping
the desired behaviors when Pods with sidecars are terminated. While original approach works better

> with truly graceful termination where consistency is more important than time taken, proposed approach

Reviewer 3: this sentence may need some work to make it easier-readable.

Reviewer 5: +1

More detailed rationale would also be worthwhile here.

The trade off between guaranteeing a per-sidecar grace period and guaranteeing a pod level grade period is one of the biggest implications to this PR (the other being the sidecar restart  behavior after main container shutdown).

Specifically, I think we should acknowledge here in the KEP that some sidecars benefit from being able to flush data to external systems during shutdown. Logging and monitoring sidecars are prime examples.  These sidecars are impacted by the decision this PR codifies.

Could we either add an entry to the alternatives considered that summarizes the idea of providing per-sidecar graceful termination periods and explains the pros, cons and the rationale for why we're going this way instead? 

It might also be worthwhile to add some use cases, both for the usage patterns that benefit from this decision and those that do not. That  would be a good place to describe how the use cases that don't benefit should cope with the design.  What would be the best practices for a logging sidecar given this design?

works for that scenario as well as a more and more popular scenario of limited time to terminate when
graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
as much as possible (think of a Spot Instances with 30 seconds notification).

> Here is the proposed approach:

Reviewer 6: Just to fully check my understanding, for a service mesh we would:

* On preStop: signal proxy to start draining (draining == encourage clients to send elsewhere with GOAWAY/`connection: close`, etc)
* On sigterm: exit asap, we know the app container is gone

?

Author: Yes

Reviewer 4: Not only are any main containers gone, but also any sidecars that started after this sidecar (e.g. put your app mesh as the first sidecar and other sidecars can rely on it and will be terminated before it).

1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
   the Pod's termination grace period countdown.
2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
3. Once the last primary container terminates, the last started sidecar container is notified by

>    sending a `SIGTERM` signal, even if `PreStop` hook hasn't completed yet.

Reviewer 5: This is a change from the [currently documented behavior](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) of PreStop: "the hook must complete before the TERM signal to stop the container can be sent".

Have we considered keeping the behavior consistent across container types (init and main)? We can still terminate based on pod termination grace period.

Reviewer 4: Good catch, that's actually how I implemented it in https://github.com/kubernetes/kubernetes/pull/120620 .

Reviewer 3: As discussed today at Sidecar WG meeting, we need to be careful to make sure the sidecar scenarios can be implemented with the minima risk. We want to give a signal at the beginning of termination AND we want another signal to notify when all main containers are gone.

As written the thinking was that sidecars will follow the general pattern:

1. On `PreStop` start termination procedures, but must function. 
   a. Logging/metrics will flush buffers and move into mode of operation with the minimal buffering
   b. Service mesh will signal proxy to start draining (draining == encourage clients to send elsewhere with GOAWAY/connection: close, etc)
2. On `SIGTERM` make the final clean ups ASAP.
  a. Logging/metrics will flush the very small buffer they have left and exit as fast as it can
  b. Service mesh will exit immediately as it is not needed any longer

The idea of sending the `SIGTERM` during the `PreStop` was that implementation-wise, sidecars will implement the main clean up logic in OreStop hook itself and will have a `SIGTERM` signal to be used to interrupt whatever `PreStop` was doing.

The alternative implementation that satisfy both scenarios, while keeping the semantic of uninterrupteable `PreStop` would be for sidecar to use `PreStop` as a notification mechanism only. So sidecar implementation may be as following:

- If sidecar implemented PreStop's clean up synchronous, the whole PreStop will run and SIGTERM will be sent to the Container after both: PreStop has completed AND all main containers are gone. There may be a gap between the PreStop completion and  SIGTERM. And there may be a situation when the sidecar extended the pod lifetime to the duration of it's PreStop execution.
- Sidecar can choose to implement PreStop as a notification mechanism. PreStop starts the termination in separate thread (let's call it `PreStopAsync`) and exits the callback immediately. The `PreStopAsync` and `SIGTERM` will behave the same as written currently in the document - `SIGTERM` can interrupt the `PreStopAsync`.

The downside of this approach is that:
1. There is a change in behavior - now there is potential gap between the `PreStop` completion and the `SIGTERM`. Authors of sidecars will need to implement container to be fully functional after `PreStop` has finished.
2. Even if PreStop is fast and only starts the `PreStopAsync`, it is still an overhead. For extra-fast Pods that needs to terminate immediately, `SIGTERM` interrupting the `PreStop` will work faster.

WIth all this said, I feel like breaking the behavior by interrupting `PreStop` sounds like a better alternative. @jpbetz @tzneal wdyt?

Reviewer: > There is a change in behavior - now there is potential gap between the PreStop completion and the SIGTERM

This is not a "change" - it's true today, just less noticeable.

I'm not seeing why we would want different semantics for this, but I am not sure I understand the proposal in full..

Reviewer 3: > > There is a change in behavior - now there is potential gap between the PreStop completion and the SIGTERM
> 
> This is not a "change" - it's true today, just less noticeable.

From implementation perspective it is pretty much immediate today. So some implementaiton might take dependency on it. After discusssion at today meeting we decided to go with it though. It will be easier for people to implement.

> I'm not seeing why we would want different semantics for this, but I am not sure I understand the proposal in full..

The main motivation here is to make sure that scenarios of spot instances with the limited graceful termination time are supported as well as we can. So we want to start termination of sidecars as soon as we can so in case of logging - they have time to flush buffers. We are talking about 15 seconds total graceful termination period here.

Ae also want to make sure sidecars are working while main containers are working, but not much longer. So we need to move the `SIGTERM` to the place that is more appropriate for sidecars.

Reviewer 2: > After discusssion at today meeting we decided to go with it though. It will be easier for people to implement.

(just to check if I understand correctly)
The sidecar container will get SIGTERM after it completes the preStop hook or the TGPS expires.
And users can implement their preStop hook in whatever way they want (synchronously or asynchronously), right?

> 4. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous

Reviewer: How long does each sidecar get between SIGTERM and SIGKILL ?  Or are you saying that it is ALL bounded by the pod's main TGPS ?  Considering that we (I) have previously written that serving apps SHOULD NOT self-exit, this means that many will receive SIGTERM and just wait for SIGKILL, leaving nothing for the sidecars.

Reviewer 3: this is fine of they will not leave anything to sidecars. With the proposal sidecars will still receive the notification to clear buffers in a form of `preStop` hook. Once preStop received they can switch to small buffer mode AND start flushing all buffers.

The alternative approach that we discussed in alpha and linked as a google doc, if the main container simply ignores SIGTERM, sidecars are missing any notifications. So often logs will be missing and termination will not be as graceful as it could have been.

   sidecar container terminates.
5. This continues until all sidecar containers have terminated, or the Pod's termination grace period
   expires. In the latter case, all containers are terminated with minimum grace period and the Pod
   will be terminated after that.

If we compare this to the initial proposal, the following behaviors are preserved:
- Sidecars should not begin termination until all primary containers have
  terminated.
- Sidecars should terminate serially and in reverse order. I.e. the first
  sidecar to initialize should be the last sidecar to terminate.

The additional benefits of this approach comparing to initial proposal:
- If graceful termination period is short, and mostly taken by the main container, the sidecar containers
  has more time to gracefully terminate, for example, clear up buffers of logging container.
- There is absolutely no change in behavior of main containers - they start graceful termination at exact
  same time as before and can utilize as much of the graceful termination period as they need. The Pod graceful
  termination period semantic also stay unchanged.

However, the following behaviors is not preserved:

> - Sidecars should continue to be restarted until all primary containers have terminated.

Reviewer 3: @howardjohn wdyt from istio perspective? sidecars will NOT be restarted while the main container is in preStop or ignoring the `SIGTERM`

Reviewer 6: Not too worried about this; it is an extreme edge case for the sidecar to crash, and since its during shutdown it doesn't seem so bad

Reviewer 5: I agree this is an edge case.  I didn't follow the motivation for not restarting them here though.  The crashes should be very rare, so wouldn't negatively impact expected shutdown duration. Is there another reason to not restart them?

Reviewer 4: Sidecar containers would have their preStop called upon pod termination beginning and we discussed that it seemed odd to to later retstart a pod after it had already received the preStop.  You couldn't reliably use the preStop as an indicator to shutdown and would be forced to wait for the SIGTERM.

Reviewer 7: @markmandel
I apologize for bringing you into this discussion.

> sidecars will NOT be restarted if crashed while the main container is in preStop or ignoring the SIGTERM

In [Agones](https://agones.dev/site/), we often set the `terminationGracePeriodSeconds` of GameServers to a higher value to prevent Pods from terminating before a game match has finished. Once a game match has finished, the main container (e.g. Unity Server) initiates a graceful termination process and invokes the [Shutdown()](https://agones.dev/site/docs/guides/client-sdks/#shutdown) method via the Agones SDK sidecar container to change the GameServer’s state to `Shutdown`. The duration of these GameServers’s `terminationGracePeriodSeconds` depends on the game’s specifications, but it can sometimes become extended to one hour or more due to multiple game matches being played by the same users. While I understand that this is an edge case, but there is a possibility that the Agones SDK sidecar container may crash. In such scenarios, we expect the sidecar container to be restarted because the main container can no longer change the state of GameServer via Agones SDK sidecar container. Alternatively, the main container cannot invoke the [Health()](https://agones.dev/site/docs/guides/client-sdks/#health) method to notify that the GameServer’s state is heathy leading to an unexpected restart by Agones controller. Could this potentially be an issue for Agones use cases involving sidecar containers? (I’m not sure but It may also present problems for [Dapr](https://docs.dapr.io/) use cases.)

Reviewer 8: Sounds like this is a resolved conversation.

(a) Thanks for pulling me in, I wasn't aware of this KEP (very excited by it! This is going to remove a bunch of hackery we have in Agones) (b) Yes, you raise a really good point -- if the sidecar did crash during the `terminationGracePeriodSeconds` period, we (Agones) would definitely want the sidecar container to restart, so that our connected SDK operations would be able to continue to operate (Health checking being one of them, but also being able to just access all the information that the SDK still provides.

So the line that I read above:

> Sidecars should continue to be restarted until all primary containers have terminated.

Sounds perfect to me.

Reviewer: IIUC this is the same argument as https://github.com/kubernetes/enhancements/pull/4183#discussion_r1341703704 ?  @markmandel can you cross-check me, I think we are saying the same thing.

Reviewer 8: Ah! Thanks for the correction. I did not read the top line. So yes - we would definitely prefer it if sidecars would continue to be restarted until the main container have terminated.

And yes, I think we are saying the same thing here. It's exactly the same issue with a logging sidecar.

- Sidecars should terminate serially and in reverse order. I.e. the first
  sidecar to initialize should be the last sidecar to terminate.

1. Using `SIGTERM` signal to notify a sidecar container when regular containers have terminated:
   - this is the same signal that is used to notify regular containers that the Pod is terminating
   - it is sent after the previous container has terminated to guarantee ordering
   - advantages:
     - it is simple and consistent with regular containers

>      - it should not break existing sidecars implemented as regular containers

Reviewer 2: FYI:
What I was trying to say in the last meeting was that
preStop hook and emitting SIGTERM are not completely separate concepts (at least in terms of their implementation).

In my POV, they are one task(container termination) and the task has consecutive works(executing the preStop hook and emitting SIGTERM) sharing the deadline(terminationGracePeriodSeconds).

My point is that it may not be consistent or easy to implement.
(just want to point out its complexity)

Author: Thanks for pointing this out @gjkim42 so what you're saying it would be easier to initiate the whole `killContainer` (hook + sigterm) for a sidecar once the previous container (or sidecar) has terminated, rather than separating their concerns?
Sounds reasonable if that is possible at all. WDYT @SergeyKanzhelev @tzneal ?

Reviewer 2: > what you're saying it would be easier to initiate the whole `killContainer` (hook + sigterm) for a sidecar once the previous container (or sidecar) has terminated, rather than separating their concerns?

Yes, I prefer that way.

If the goal is to implement termination ordering while also providing a termination graceful period for sidecar containers without introducing any new field. I'm okay with the current way(separate prestop and SIGTERM). However, it may not be as simple as we thought.

Reviewer 2: Yes, it could work.

From the user's perspective, I think of a `preStop` as a part of the container termination, so separating it looks a bit complicated. It looks quite exceptional.

---
> - Sidecars should not begin termination until all primary containers have terminated.
>   - Implicit in this is that sidecars should continue to be restarted until all primary containers have terminated.

One other question is what happens if the sidecar container crashes during the regular container's termination?
If we want to ignore the crashed sidecar container, it is just ok.
else if we want to restart the sidecar container, how can we handle the preStop hook?

Reviewer 2: > I may be misunderstanding, but your suggestion is that pre-stop and sigterm are essentially linked events, and a sidecar wouldn't get the pre-stop until all main containers and all sidecars that started after it have stopped?

Yes.

My point was that if we choose early pre-stop, even after executing the pre-stop we may have to restart the sidecar container (because it crashes after executing the pre-stop hook). It looks weird to me.

Reviewer 4: > My point was that if we choose early pre-stop, even after executing the pre-stop we may have to restart the sidecar container (because it can crash after executing the pre-stop hook). It looks weird to me.

Reviewer 2: > The advantage to the early pre-stop is that for the case of main containers that consume the full grace period, sidecars are at least aware that they'll be shutting down for a period of time before they just get a SIGKILL when the full grace period is expired.

Yes, I agree that we have to provide some way for a sidecar to be aware of its termination.

However, it would be better not to use pre-stop but to add or borrow the dedicated termination period for the sidecars to make them aware of its termination. (And yes it may need an extra field but I think it is necessary.)

This might just be my own opinion, so I'd appreciate hearing more thoughts.

Reviewer 2: > getting zero notice before a SIGKILL if the main containers don't behave well.

https://github.com/kubernetes/kubernetes/blob/58ce7342230a943b637ff3d129eb3774d5a68cbb/pkg/kubelet/kuberuntime/kuberuntime_container.go#L741-L744

I know it is not enough but we have assigned `minimumGracePeriod` (2s).

### Termination of containers

In Alpha sidecar containers will be terminated as regular containers. No special
or additional signals will be supported.

In Beta we have thought about introducing additional termination grace period fields
to manage termination duration
([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
and leverage these fields to add reverse order termination of sidecar containers
after the primary containers terminate.

However, we decided on an alternative that doesn't require additional fields or hooks while keeping
the desired behaviors when Pods with sidecars are terminated. While original approach works better
with truly graceful termination where consistency is more important than time taken, proposed approach
works for that scenario as well as a more and more popular scenario of limited time to terminate when
graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
as much as possible (think of a Spot Instances with 30 seconds notification).

Here is the proposed approach:

> 1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating

Reviewer: I want to float for consideration: do we want a different hook?  We could define `preStopPod` with clear semantics that are meaningful.  If we go beta with `preStop` the way this is specc'ed, and we're wrong, we have to add `preStopContainer` or something which is less obvious.  

The `preStop` docs also say that it is called in case of a livenessProbe failure.

Reviewer 3: liveness probe is a very good point. I totally forgot about this.

So if people implement `preStop` as a "signal" and initiate flushing of buffers, while in no-buffer mode, than buffers will be lost.

If sidecar is implemented as full flush in preStop, while in no-buffer mode, than liveness behavior will be OK. The problem will be that the sidecar is potentially delaying the Pod termination for the duration of it's preStop.

So if buffers are less important on Pod termination while they are very important on liveness probe failure, than we have a problem.

Reviewer 3: I am hesitant to add a new hook that only has meaning for the sidecars. Do you think `preStopPod ` may be useful for regular containers?

Reviewer: I think the fact that `livenessProbe` uses `preStop` _requires_ a different hook.

Consider a logging sidecar.
* it starts up
* call `postStart`
* app containers start writing logs
* logger fails livenessProbe
* call `preStop` (which, now that I think of it, is kind of weird)
* restart logger
* Pod is deleted
* call `preStop`
* app container stays alive until SIGKILL
* logger is killed with 0 grace

Where in there should it have started aggressively flushing buffers?

I went back to the diagrams I started earlier this year and realized I don't think I shared them publicly: https://docs.google.com/presentation/d/1e-qJWe6He2qjt0PFiZtocqP8VbkrwKAI9B4X72L0bGk/edit?usp=sharing&resourcekey=0-N77W7Q5UHuN5dFtekjmAIA - I have not added sidecars here, but looking at this, `preStop` is pretty clearly ambiguous.
 
If we can't change the definition of `preStop`, we need something like `preStopPod` or `podDeleted`.  Adding API at this point means either a second feature gate or this stays in Alpha.

Thought experiment: if we consider SIGTERM as just another form of lifecycle hook, what would you name it?  `stop` ?  That's a little confusing - one might assume that `preStop` is always followed by `stop`, but it's not.  Maybe `eol` or `terminated`?  That way you could say that a pod being deleted will see:

* deletionTimestamp is set
* all containers get `podDeleted` hook
* app containers get `preStop`
* app containers get `terminated` (SIGTERM)
* wait up to TGPS
* app containers get `SIGKILL`
* if there is time left
  * sidecars get `preStop`
  * sidecars get `terminated` (SIGTERM)
  * wait up to remainder of TGPS
  * sidecars get SIGKILL

In fact we could literally model `terminated` as a hook in `lifecycle`, which defaults to `signal: { name: "SIGTERM" }` (solving a different long-standing issue :) -- but not in this KEP.

Unresolved: do we send preStop to all sidecars at once or one at a time in reverse order?  If we were to later add a per-container grace period, I think we would say one at a time, right?

Reviewer 3: In my mind it's simpler:

- `preStop` == prepare to be terminated. You may still be needed though.
- `SIGTERM` == we strongly insist you exit on your own, it's time now and nobody needs you.
- `SIGKILL` == ok, we warned you, done.

In this sense, liveness is not much different than pod completion. On `preStop` you start flushing aggressively, but can still accept logs. Prepared to be deleted. If you want to make liveness `preStop` to be faster, you can always set `terminationGracePeriodSeconds` override on the liveness probe.

Thinking more about it, new lifecycle hook will just complicate things

Reviewer: `preStop` on a liveness failure is a totally different thing from a deleted pod.  There's no need to flush buffers on a liveness failure.  In fact, I think you DON'T WANT TO - you want to do as little work as possible and die, so you can be restarted, right?

If I have a pod who fails liveness, we call preStop.  Do we also send SIGTERM, wait, SIGKILL?  I don't think so (except the new probe-level setting, which I hate).  The docs make it sound like we do, but I didn't see SIGTERM in that case and my diagrams do not have it.

If we overload preStop this way AND go beta, we can't easily take it back (not impossible but hard).    Are you willing to put this all behind a new alpha gate?

Reviewer 3: I think diagram is wrong. We are calling a regular `killContainer` (https://github.com/kubernetes/kubernetes/blob/622509830c1038535e539f7d364f5cd7c3b38791/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1093).

> `preStop` on a liveness failure is a totally different thing from a deleted pod. There's no need to flush buffers on a liveness failure. In fact, I think you DON'T WANT TO - you want to do as little work as possible and die, so you can be restarted, right?

This is exactly the reason the override was introduced. So liveness probe's preStop can be cut short.

Depending on how container designed, you may expect that the buffers need flushing after liveness probe failed. But in general I agree this may not be very reliable.

> If I have a pod who fails liveness, we call preStop. Do we also send SIGTERM, wait, SIGKILL? I don't think so (except the new probe-level setting, which I hate). The docs make it sound like we do, but I didn't see SIGTERM in that case and my diagrams do not have it.

Yes, those will be sent.

> If we overload preStop this way AND go beta, we can't easily take it back (not impossible but hard). Are you willing to put this all behind a new alpha gate?

The proposal is to keep the behavior that exists today.

I agree with your reasoning. It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this. One way to do it in future may be to pass additional arguments to the lifecycle hook to distinguish signals. Fixing is one way to sidecars and another way to regular containers will be the worst of both worlds

Reviewer: > I think diagram is wrong.

I went and proved you right.  I updated the drawings.

> This is exactly the reason the override was introduced. So liveness probe's preStop can be cut short.

We never should have called preStop on liveness fail.  It should have been different.

> The proposal is to keep the behavior that exists today.

I barely count alpha as a commitment.  Once we do this in beta, it's a much stronger statement of "we considered it and this is how we want it".  What that means is `preStop` == `podDelete`.

So rewriting my timeline:

* `deletionTimestamp` is set
* all containers get `preStop` hook
* app containers get `terminate` (implicitly == SIGTERM)
* wait up to TGPS
* app containers get SIGKILL
* iff there is time left
  * sidecars get `terminate` (implicitly == SIGTERM)
  * wait up to remainder of TGPS
  * sidecars get SIGKILL

> It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this. 

I don't see what you mean?  If we spec `preStop` as "this container is being stopped" and `podDelete` as "the pod is being stopped" (with the guarantees that `podDelete` does not happen in case of liveness fail and `podDelete` always happens before `preStop` in case of pod termination), the only containers who need to adapt are sidecars, and they already need to adapt.

What's the reason to not add a specific `podDelete` hook?  Is it just to avoid another gate and/or alpha (which is somewhat valid)?  Or is there actually a semantic argument?

In your model: do we send `preStop` to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?

Reviewer 3: > > The proposal is to keep the behavior that exists today.
> 
> I barely count alpha as a commitment. Once we do this in beta, it's a much stronger statement of "we considered it and this is how we want it". What that means is `preStop` == `podDelete`.

What I meant here is that we should be keeping the behavior of calling `preStop` on liveness probe failure for sidecars similarly how it is done on regular container is what I mean by existing behavior. Changing this semantic should ideally be done consistently across regular and sidecar containers.

> > It feel more natural to have a signal differentiated. But it is not a problem scoped to sidecars, all containers needs changing if we decide to improve this.
> 
> I don't see what you mean? If we spec `preStop` as "this container is being stopped" and `podDelete` as "the pod is being stopped" (with the guarantees that `podDelete` does not happen in case of liveness fail and `podDelete` always happens before `preStop` in case of pod termination), the only containers who need to adapt are sidecars, and they already need to adapt.

I read the semantic of `preStop` as a signal to start packing. Container will be stopped soon, but we will do our best to wait for you "pack your things". Not terminating your yet, so please keep the readiness probe up if you can serve things. But letting you know.

> What's the reason to not add a specific `podDelete` hook? Is it just to avoid another gate and/or alpha (which is somewhat valid)? Or is there actually a semantic argument?

For me it is a semantic argument. I want to keep sidecars behavior aligned with other containers in terms of lifecycle hooks. And I think sidecars fit perfectly into the model today to enable most of termination scenarios. 

> In your model: do we send `preStop` to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?

All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequencial, but non-blocking, correct?

Reviewer: > What I meant here is that we should be keeping the behavior of calling preStop on liveness probe failure for sidecars similarly how it is done on regular container

Oh.  I agree with that.  IMO that mnakes it more obvious that `preStop` and `podDelete` are different things.  I don't think we can change it.

> I read the semantic of preStop as a signal to start packing. Container will be stopped soon

Yes, I think that's true in either approach.

>> What's the reason to not add a specific podDelete hook? Is it just to avoid another gate and/or alpha (which is somewhat valid)? Or is there actually a semantic argument?
> 
> For me it is a semantic argument. I want to keep sidecars behavior aligned with other containers in terms of lifecycle hooks. And I think sidecars fit perfectly into the model today to enable most of termination scenarios.

It fits fine, if you squint a little.  You say "aligned with other containers " but you are pinning that on one interpretation.  

Today, `preStop` means:
* the container is going down soon (e.g. liveness)
* the pod is going down soon (e.g. deleted)
* start your end-of-life

This is fine because we have no sequence.  As soon as we add a sequence, we need to decide which of those `preStop` means.  Because of the liveness probe semantic, it CAN'T mean "the pod is going down soon".   You are saying it should mean "start your end-of-life" and pods don't get to know why.  I am saying it should mean "the container is going down soon" and a different signal should mean "the pod is going down soon".  Both work, I think.  Your path is simpler and informative.  Other path is richer but more complex.

Is that a fair breakdown?  

>> In your model: do we send preStop to all sidecars at once or one at a time in reverse order? If we were to later add a per-container grace period, I think we would say one at a time, right?
>
> All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequencial, but non-blocking, correct?

Can we spec that preStop and SIGTERM are always delivered in reverse-startup order, though there may be almost zero time between them?  That leaves room for adding grace periods later if we need.

Reviewer 3: > This is fine because we have no sequence. As soon as we add a sequence, we need to decide which of those `preStop` means. Because of the liveness probe semantic, it CAN'T mean "the pod is going down soon". You are saying it should mean "start your end-of-life" and pods don't get to know why. I am saying it should mean "the container is going down soon" and a different signal should mean "the pod is going down soon". Both work, I think. Your path is simpler and informative. Other path is richer but more complex.
> 
> Is that a fair breakdown?

Yes, it is fair. More signal is better for more advanced scenarios. But complicates documentation and implementation. 

Also we may enrich things by passing more arguments to the lifecycle hooks in future in form of env vars or http headers. It may be better than adding more hooks or change the pod specs.

> Can we spec that preStop and SIGTERM are always delivered in reverse-startup order, though there may be almost zero time between them? That leaves room for adding grace periods later if we need.

Yes, this is a great idea. If we have a definitive order for the regular containers we can spec it as well.

Reviewer 4: > All sidecars at once. I do not have any ordering preferences. @tzneal implementaiton-wise is there any ordering for the regular containers today? I believe it will be sequential, but non-blocking, correct?

No, there's no ordering for regular containers. It's a go-routine to kill each container, and waiting on them all to finish.  https://github.com/kubernetes/kubernetes/blob/6a84edb2cee68fc60d3f4bd15eb461a184303236/pkg/kubelet/kuberuntime/kuberuntime_container.go#L768

Reviewer: I don't really care about app containers - more concerned about sidecars.

   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
   the Pod's termination grace period countdown.

> 2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.

Reviewer: Is this a feature ("we don't want them to restart") or a lack of feature ("we would prefer they restart but it's hard") ?

We still (TTBOMK) don't have a good state diagram for pod lifecycle.  This seems like it could be a problem.

E.g. if I have a 60m TGPS and my logging sidecar dies, no more logs?

Reviewer 3: yes, this is what will happen. The problem is that it is close to impossible to distinguish desired termination vs. crash.

Reviewer: > The problem is that it is close to impossible to distinguish desired termination vs. crash.

I don't understand what you mean.  If `deletionTimestamp` is set it's a termination, if not it's a crash?

Reviewer 3: I am talking about a sidecar. Once deletionTimestamp is set, we call `preStop`. Sidecar has an option to exit on a spot. And restarting it after this may not be what user will desire. However we cannot distinguish if container exit by itself or crashed.

Reviewer: Make it `OnFailure` ?  Frankly, "restartPolicy=Always" was probably a mistake.  We should have only had "Never" and "OnFailure".

Reviewer: xref https://github.com/kubernetes/enhancements/pull/4183#discussion_r1342069396 - I think it's the same point.

Reviewer 8: Bringing my comment up here. Same issue on [Agones](https://agones.dev/) we run a sidecar SDK server that is connected to over localhost, ideally we would want the sidecar to restart while within a TGPS timeframe.

> The problem is that it is close to impossible to distinguish desired termination vs. crash.

Probably a naive question (with no idea how kubelet etc work), but do exit codes help here?

Reviewer 9: @markmandel I think `OnFailure` is the exit code based solution - and I like the idea that sidecars restart by default `OnFailure`.

Reviewer 3: restarting past the deletion timestamp on failure sounds like a good addition. However it complicates things a bit - do we call `preStop` right after start on every sidecar restart?

Reviewer 8: ...and TIL what `OnFailure` is. Honestly, wasn't even aware that was a thing! 😄 

In which case +1 on a restart policy `OnFailure` as the option during TGPS timeframe. That seems ideal to me, and also gives enough control to the end user.

Ignore if already been covered, but also floating the idea of giving the end user the option of what restart policy they want during TGPS for a sidecar? I could also understand if that's way too much complexity for container/pod lifecycle.

Reviewer 3: > Ignore if already been covered, but also floating the idea of giving the end user the option of what restart policy they want during TGPS for a sidecar? I could also understand if that's way too much complexity for container/pod lifecycle.

we don't support this for regular containers. I'd say if we need it, we will need to do a bigger change that is not necessarily bound to sidecars.

We want to preserve the behavior of TGPS - it should be respected no matter whether sidecars are there or not. And we also think that if we will start taking parts of TGPS for sidecars exclusively, than back compat will break causing more problems.

Reviewer 8: That's fair. Just figured I'd float the idea to see if it stuck in any way.

Reviewer: > Once deletionTimestamp is set, we call preStop. Sidecar has an option to exit on a spot ... we cannot distinguish if container exit by itself or crashed.

I disagree - `preStop` is NOT a signal that you should exit, it's a signal that your time to exit is imminent.  Hence "pre" in the name.  If we call `preStop` and then you exit, we _should_ restart you EVEN IF YOU EXIT(0).  We did not call SIGTERM yet, so you are still subject to restarts.  We don't have a "terminating" state, per se, but deletionTimestamp is insufficient.

Reviewer 7: I agree with @thockin. I'm not sure about the specific use case for terminating a process on its own within the PreStop Hook. Is this related to "Abbreviated termination" scenario? If so, should the sidecar container exit the PreStop Hook as early as possible and just let SIGTERM to terminate the process?

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> - <test>: <link to test coverage> FIXME to fill

Author: @tzneal @gjkim42 can you help me here?

Reviewer 2: Maybe...

`No integration tests are planned. We'll cover this with e2e_node tests.`

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> - <test>: <link to test coverage> FIXME to fill

Author: @tzneal @gjkim42 here too?

Reviewer 2: adding test lists and their failure rate link to each subsection should be enough.

`ginkgo run -v --dry-run --focus=SidecarContainers ./test/e2e_node/`
this command should be helpful.
```
$ ginkgo run -v --dry-run --focus=SidecarContainers ./test/e2e_node/
...
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container runs continuously should 
keep running a Pod continuously and produce log                                                                                                                                                             
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1889                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------                                                                                                                                                                              
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container runs continuously should 
not restart a restartable init container                                                                                                                                                                    
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1901                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------                                                                                                                                                                              
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container runs continuously should 
start a regular container                                                                                                                                                                                   
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1905                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------                                                                                                                                                                              
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container fails to start because of
 a bad image should continuously run Pod keeping it Pending and produce log                                                                                                                                 
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1948                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------                                                                                                                                                                              
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container fails to start because of
 a bad image should not start a restartable init container                                                                                                                                                  
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1960                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------                                                                                                                                                                              
[sig-node] [NodeAlphaFeature:SidecarContainers] Containers Lifecycle  when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container fails to start because of
 a bad image should not start a regular container                                                                                                                                                           
/home/gjkim/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1963                                                                                                                         
• [0.000 seconds]                                                                                                                                                                                           
------------------------------
...
```
